design/task-management-api.md

```diff
@@ -0,0 +1,385 @@
+# Task management API: requirements
+
+Multiple subsystems require reliable long-lived distributed operations.  We shall support them
+with a task queue subsystem.  Concepts of the API are intended to parallel concepts in the
+[Celery][celery-api] and/or [Machinery][machinery-api] APIs.  We do not use either of them to
+reduce the number of required dependencies.  A future version may support external queues to
+achieve better performance or robustness, at the price of increased ops or cost.  These will
+likely be queues such as Kafka or SQS rather than full Celery or Machinery.
+
+In practice this generally means providing APIs similar to those of Machinery (which is more
+Go-like than Celery) for constructing task flows and for registering workers.
+
+In particular:
+1. We provide similar concepts for building task flows as do existing
+   task queues.
+1. We use similar terminology.
+1. We do *not* require the entire API of an existing task queue.
+1. We do *not* use the verbs or API calls of an existing task queue.
+
+This API definition comes with implementation sketches for how we to use these APIs to implement
+the branch export and concurrent dedupe "user" stories.
+
+## API
+
+### Concepts
+
+#### Tasks
+
+A task is the basic atom of task management.  It represents a single unit of work to
+perform, and can succeed or fail.  Tasks may be retried on failure, so _executing a task
+must be idempotent_.
+
+Tasks connect to application code via an action and a body.  The _action_ identifies the
+operation to complete these task.  Examples of actions can include "copy a file", "delete a
+path", "report success".  It is essentially the _name_ of a procedure to perform at some future
+time.  The _body_ of a task gives information necessary to configure the specific task.
+Examples of bodies can include "source file path X, destination path Z" (for a copy task), "path
+Z" (for a delete task), or "date started, number of objects and a message" (for a report task).
+It essentially holds the _parameters_ the action uses to perform the task.
+
+Tasks include these attributes:
+- `Id`: a unique identifier for the task.  Use a known-unique substring in the identifier
+  (e.g. a UUID or [nanoid][nanoid]) to avoid collisions, or a well-known identifier to ensure
+  only one task of a type can exist.
+- `Action`: the type of action to perform for this task.  Workers pick tasks to perform and the
+  actions to perform on them according to this field.
+- `Body`: a description of parameters for this text.  E.g. in a "copy file" task the body might
+  specify source key, ETag and destination key.
+- `StatusCode`: the internally-used state of the task in its lifecycle, see [life of a
+  task](#life-of-a-task) below.
+- `Status`: a textual description of the current status, generated by application code.
+- `ToSignal`: an array of task IDs that cannot start before this task ends, and will therefore
```

Comment:
How do you avoid cycles?

Reply:
Easy -- you don't.  I don't believe it is worth the effort of solving this problem, it is a logic error in the calling code to do this.  The calling code might equally _expect_ a task to have finished when it hasn't (a missing dependency), or get stuck (a "dynamic halt" aka infinite loop).  Topologically sorting dependencies requires about as much work as generating them.

That said, we _can_ produce utility functions to generate chains and chords.  And I think I _should_.